ferret/speechxai_utils.py

```diff
@@ -7,6 +7,81 @@
 import torch
 from datasets import Dataset
 from transformers import Wav2Vec2ForSequenceClassification, Wav2Vec2FeatureExtractor
+import librosa
+from typing import Union
+from explainers.explanation_speech.utils_removal import transcribe_audio
+
+
+class FerretAudio:
+    """
+    Internal class to handle audio data. We force signal to 1) mono, 2) a sampling rate of 16000, 3) np.float32 (i.e., 4 bytes to represent each sample).
+    We infer the native sampling rate using librosa.
+    """
+
+    def __init__(
+        self,
+        audio_path_or_array: Union[str, np.ndarray],
+        native_sr: int = None,
+        model_helper=None,
+    ):
+        self.target_sr = 16000
+        self.native_sr = native_sr
+        self.audio_path_or_array = audio_path_or_array
+        self.model_helper = model_helper
+        self._transcribe = None
+
+        if isinstance(audio_path_or_array, str):
+            self.native_sr = librosa.get_samplerate(audio_path_or_array)
+            self.array, self.sample_rate = librosa.load(
+                audio_path_or_array, sr=self.target_sr, dtype=np.float32
+            )
+
+        elif isinstance(audio_path_or_array, np.ndarray):
+            if native_sr is None:
+                raise ValueError(
+                    "If audio is provided as a numpy array, native_sr must be provided"
+                )
+            self.array, self.sample_rate = librosa.resample(
+                audio_path_or_array, self.native_sr, self.target_sr
+            )
+
+    @property
+    def is_normalized(self) -> bool:
+        """Check if the array is already normalized."""
+        return np.max(np.abs(self.array)) <= 1.0
+    
+    @property
+    def normalized_array(self) -> np.ndarray:
+        if not self.is_normalized:
+            return self.array / 32768.0
+        else:
+            return self.array
+
+    @property
+    def transcribe(self):
```

Comment:
I would have called this `transcription` as it's used as an attribute - but I'm aware I'm being pedantic here ahah!